### PR TITLE
Fix/nodes run timeout propagation

### PR DESCRIPTION
**Pull request by danedane**
Summary

• Problem: nodes run (system.run) kills any node command taking >30s because callGatewayTool uses gatewayOpts.timeoutMs (default 30s) regardless of commandTimeoutMs or invokeTimeoutMs values.
• Why it matters: Makes nodes run unusable for any real workload (builds, coding agents, long scripts) on paired nodes.
• What changed: Compute runGatewayTimeout from invokeTimeoutMs or commandTimeoutMs + 15s, propagate it to both the websocket call options and the message-level timeoutMs.
• What did NOT change (scope boundary): No schema changes, no gateway-side changes, no new parameters. Only the agent-side timeout propagation logic in nodes-tool.ts.

Change Type (select all)

• [x] Bug fix
• [ ] Feature
• [ ] Refactor
• [ ] Docs
• [ ] Security hardening
• [ ] Chore/infra

Scope (select all touched areas)

• [ ] Gateway / orchestration
• [x] Skills / tool execution
• [ ] Auth / tokens
• [ ] Memory / storage
• [ ] Integrations
• [ ] API / contracts
• [ ] UI / DX
• [ ] CI/CD / infra

Linked Issue/PR

• Closes #
• Related #

User-visible / Behavior Changes

• nodes run commands with commandTimeoutMs or invokeTimeoutMs set will now correctly wait for the node to complete instead of timing out at 30s.
• No new parameters or config changes. Existing behavior for commands <30s is unchanged.

Security Impact (required)

• New permissions/capabilities? No
• Secrets/tokens handling changed? No
• New/changed network calls? No
• Command/tool execution surface changed? No
• Data access scope changed? No

Repro + Verification

Environment

• OS: WSL2 (Ubuntu) + Windows 11 paired node
• Runtime/container: Node v22.22.0, OpenClaw v2026.3.8
• Model/provider: github-copilot/claude-opus-4.6
• Integration/channel: Telegram
• Relevant config: Default gateway config, Windows companion node paired via LAN

Steps

Pair a Windows node with OpenClaw gateway
Run nodes run with a command that takes >30s (e.g., ping -n 35 127.0.0.1) and commandTimeoutMs: 60000
Observe timeout behavior
Expected

• Command completes after ~35s and returns stdout with all 35 ping replies

Actual

• Before fix: gateway timeout after 30000ms error at 30s
• After fix: Command completes successfully (34.6s), returns full output

Evidence

• [x] Trace/log snippets

Before fix:

{"status":"error","error":"gateway timeout after 30000ms"}

After fix (monkey-patched dist):

{"stdout":"...35 pings...","exitCode":0,"timedOut":false,"durationMs":34605}

Also tested GitHub Copilot CLI (74s run) successfully via Windows node after fix.

Human Verification (required)

• Verified scenarios: 35-ping test (34.6s), Copilot CLI codebase analysis (74s), both via Windows node
• Edge cases checked: No invokeTimeoutMs set (falls back to commandTimeoutMs + 15s); neither set (preserves default 30s behavior)
• What you did not verify: Approval retry path (tested only the first-attempt path); nodes invoke action (separate code path with same issue, not addressed here)

Review Conversations

• [x] I replied to or resolved every bot review conversation I addressed in this PR.
• [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

Compatibility / Migration

• Backward compatible? Yes
• Config/env changes? No
• Migration needed? No

Failure Recovery (if this breaks)

• How to disable/revert this change quickly: Revert the single commit; timeout falls back to default 30s behavior.
• Files/config to restore: src/agents/tools/nodes-tool.ts
• Known bad symptoms reviewers should watch for:

If runGatewayTimeout computes an unexpectedly large value, gateway websocket connections could stay open longer than intended. Mitigated by only deriving from user-provided timeout params.

Risks and Mitigations

• Risk: Gateway websocket connections held open longer when large commandTimeoutMs values are passed.

• Mitigation: Timeout is derived directly from user-specified params — no unbounded defaults. If neither commandTimeoutMs nor invokeTimeoutMs is set, behavior is unchanged (30s default).

### CODE REVIEW

File: src/agents/tools/nodes-tool.ts [hunk 1]
```diff
@@ -638,6 +638,20 @@ export function createNodesTool(options?: {
...
+            // invokeTimeoutMs takes priority; fall back to commandTimeoutMs + 15s buffer.
+            // Without this, callGatewayTool uses gatewayOpts.timeoutMs (default 30s),
+            // which kills any node command that takes longer than 30 seconds.
+            const derivedTimeout =
+              invokeTimeoutMs ?? (commandTimeoutMs ? commandTimeoutMs + 15_000 : undefined);
+            // Clamp to at least the default gateway timeout floor (30s) so that
+            // small commandTimeoutMs values don't shrink below the time needed
+            // for node wake/reconnect flows.
+            const runGatewayTimeout =
+              derivedTimeout !== undefined ? Math.max(derivedTimeout, 30_000) : undefined;
```
[the remark this answers is not in the document]
Comment (danedane) [reply src/agents/tools/nodes-tool.ts]: Addressed in 9af7e20 - invokeTimeoutMS is now honored as-is (no floor clamping). The 30s Math.max floor only applies to the commandTimeoutMS-derived fallback path, so callers relying on short invoke timeouts for fast failure/retry wont be affected.

File: src/agents/tools/nodes-tool.ts [hunk 2]
```diff
@@ -638,6 +638,21 @@ export function createNodesTool(options?: {
...
+            // invokeTimeoutMs takes priority; fall back to commandTimeoutMs + 15s buffer.
+            // Without this, callGatewayTool uses gatewayOpts.timeoutMs (default 30s),
+            // which kills any node command that takes longer than 30 seconds.
+            // When invokeTimeoutMs is explicitly set, honor it as-is (callers may
+            // want short timeouts for fast failure/retry). Only apply the 30s floor
+            // to the commandTimeoutMs-derived fallback, where shrinking below the
+            // default could break node wake/reconnect flows.
+            const runGatewayTimeout = invokeTimeoutMs
+              ?? (commandTimeoutMs
+                ? Math.max(commandTimeoutMs + 15_000, 30_000)
```
[the remark this answers is not in the document]
Comment (danedane) [reply src/agents/tools/nodes-tool.ts]: adopted Greptile's suggestion. runGatewayTimeout now takes Math.max(derivedTimeout, commandTimeoutMS + 15s, 30_000), so the gateway timeout always covers the command execution time regardless of which parameter combination is used.

**Comment by steipete** [comment]
Closing this as implemented after Codex review.

Current `main` already supersedes this PR: the `nodes run` path it patches no longer exists, `system.run` is intentionally blocked from `nodes invoke`, and node shell execution now goes through `exec host=node`, which derives and applies a longer gateway timeout for `node.invoke` calls.

What I checked:

- **`nodes run` removed from agent tool:** [`src/agents/tools/nodes-tool.ts:19`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/tools/nodes-tool.ts#L19) no longer includes a `run` action, and the default branch at [`src/agents/tools/nodes-tool.ts:284`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/tools/nodes-tool.ts#L284) rejects unknown actions. ([`src/agents/tools/nodes-tool.ts:19`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/tools/nodes-tool.ts#L19), [564f820efa2b](https://github.com/openclaw/openclaw/commit/564f820efa2b))
- **`system.run` blocked on node invoke:** [`src/agents/tools/nodes-tool-commands.ts:10`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/tools/nodes-tool-commands.ts#L10) and [`src/agents/tools/nodes-tool-commands.ts:120`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/tools/nodes-tool-commands.ts#L120) reserve `system.run` / `system.run.prepare` for shell execution via `exec host=node`, so the PR's old `nodes-tool.ts` `system.run` path is gone. ([`src/agents/tools/nodes-tool-commands.ts:10`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/tools/nodes-tool-commands.ts#L10), [564f820efa2b](https://github.com/openclaw/openclaw/commit/564f820efa2b))
- **Replacement exec path propagates timeout:** [`src/agents/bash-tools.exec-host-node.ts:202`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/bash-tools.exec-host-node.ts#L202) derives `invokeTimeoutMs` from the requested exec timeout, [`src/agents/bash-tools.exec-host-node.ts:222`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/bash-tools.exec-host-node.ts#L222) forwards the command timeout into `system.run`, and [`src/agents/bash-tools.exec-host-node.ts:423`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/bash-tools.exec-host-node.ts#L423) uses that derived timeout for the gateway `node.invoke` call. ([`src/agents/bash-tools.exec-host-node.ts:202`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/bash-tools.exec-host-node.ts#L202), [564f820efa2b](https://github.com/openclaw/openclaw/commit/564f820efa2b))
- **Docs and tests confirm old interface is retired:** Docs now direct users to `exec host=node` instead of `openclaw nodes run` at [`docs/cli/nodes.md:62`](https://docs.openclaw.ai/cli/nodes), while tests assert the removed wrapper and blocked `system.run` path at [`src/cli/nodes-cli.coverage.test.ts:110`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/cli/nodes-cli.coverage.test.ts#L110), [`src/agents/tools/nodes-tool.test.ts:165`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/tools/nodes-tool.test.ts#L165), and [`src/agents/tools/nodes-tool.test.ts:312`](https://github.com/openclaw/openclaw/blob/564f820efa2b/src/agents/tools/nodes-tool.test.ts#L312). ([`docs/cli/nodes.md:62`](https://docs.openclaw.ai/cli/nodes), [564f820efa2b](https://github.com/openclaw/openclaw/commit/564f820efa2b))
- **Shipped release also contains superseding structure:** The release tag `v2026.4.22` (`00bd2cf7a376f1fba26291c6c4766f1f15cbdfa5`) shows the same `nodes` tool shape and `exec host=node` timeout logic, so this PR is already obsolete in shipped code as well as on `main`. ([`src/agents/bash-tools.exec-host-node.ts:202`](https://github.com/openclaw/openclaw/blob/00bd2cf7a376/src/agents/bash-tools.exec-host-node.ts#L202), [00bd2cf7a376](https://github.com/openclaw/openclaw/commit/00bd2cf7a376))

So I’m closing this as already implemented rather than keeping a duplicate issue open.

Review notes: reviewed against [564f820efa2b](https://github.com/openclaw/openclaw/commit/564f820efa2bb81ae49b8f55a93e41492ad31917); fix evidence: commit [564f820efa2b](https://github.com/openclaw/openclaw/commit/564f820efa2bb81ae49b8f55a93e41492ad31917).
